wagtail/images/blocks.py

```diff
@@ -51,3 +57,156 @@ def htmldiff(self):
+    """
+
+    image = ImageChooserBlock(required=True)
+    decorative = BooleanBlock(
+        default=False, required=False, label=_("Image is decorative")
+    )
+    alt_text = CharBlock(required=False, label=_("Alt text"))
+
+    def get_searchable_content(self, value):
+        return super().get_searchable_content(value)
+
+    def _struct_value_to_image(self, struct_value):
+        image = struct_value.get("image")
+        decorative = struct_value.get("decorative")
+        if image:
+            # If the image is decorative, set alt_text to an empty string
+            image.contextual_alt_text = (
+                "" if decorative else struct_value.get("alt_text")
+            )
+            image.decorative = decorative
+        return image
+
+    def to_python(self, value):
+        # For backward compatibility with ImageChooserBlock
+        if isinstance(value, int):
+            image = self.child_blocks["image"].to_python(value)
+            struct_value = {"image": image, "decorative": False, "alt_text": None}
+        else:
+            struct_value = super().to_python(value)
+        return self._struct_value_to_image(struct_value)
+
+    def bulk_to_python(self, values):
+        struct_values = []
+        for value in values:
+            # For backward compatibility with ImageChooserBlock
+            if isinstance(value, int):
+                image_values = self.child_blocks["image"].bulk_to_python(values)
+
+                for image in image_values:
+                    struct_value = {
+                        "image": image,
+                        "decorative": False,
+                        "alt_text": None,
+                    }
+                    struct_values.append(struct_value)
+
+                break  # no need for further execution after bulk_to_python has been called.
+
+            else:
+                struct_values = super().bulk_to_python(values)
+                break  # no need for further execution after bulk_to_python has been called.
+
+        return [
+            self._struct_value_to_image(struct_value) for struct_value in struct_values
+        ]
+
+    def value_from_datadict(self, data, files, prefix):
+        struct_value = super().value_from_datadict(data, files, prefix)
+        return self._struct_value_to_image(struct_value)
+
+    def clean(self, value):
+        if value is None:
+            raise StructBlockValidationError(
+                block_errors={
+                    "image": ValidationError(
+                        _("Expected an image instance, got nothing")
+                    )
+                }
+            )
+
+        if not isinstance(value, AbstractImage):
+            raise StructBlockValidationError(
+                block_errors={
+                    "image": ValidationError(
+                        _("Expected an image instance, got %(value)s")
+                        % {"value": value}
+                    )
```

Comment:
Is there a way to trigger this code path? I would like to see what this outputs.

If it is rarely executed; would a raw (untranslated) ValueError for the developer be better perhaps? Now we introduce an additional translatable string that is rarely used.

Reply:
It's almost never executed. To be frank, I'm not sure how possible it is for an image that isn't an instance of "abstractImage" to be uploaded. Perhaps it's a bit unnecessary and I can remove it. What do you think? Or should I leave it as is?

Reply:
This `clean` method will need wider reworking to support `required=False`.